### PR TITLE
[READY]Circuitry Copyright Agreement Of 2018 (make circuits id-lockable and sealable)

### PR DESCRIPTION
**Pull request by Shdorsh**
[Changelogs]: # Chuckster Corp and Nanotrasen have come to an agreement that the user of their products should have the ability to protect their designs as their own intellectual property. As such, the new printers were fitted with blueprints of permanently sealable assemblies in order to prevent tampering and theft of design. Due to a collaboration, the id-card system has been expanded to be used for locking an assembly's identity and preventing the use of analyzers on them. The use of third party analyzers and modified analyzers who can break this encryption are against the terms of service.

:cl: Shdorsh
tweak: circuits can be sealed so no one can tamper with their insides directly. Unsealing has a 50% probability to remove circuit connections.
tweak: debuggers have a special id lock setting now, that can be used to prevent analyzing assemblies.
tweak: making printers id lockable so they only print out id locked circuits. also, making the printers look slightly better.
/:cl:

[why]: Make it so that greyshits don't steal other people's design and have to come up with their own. Also, make circuits more secure to use so people can show off their best without the constant fear of the circuit analyzer.

### CODE REVIEW

**Comment by Carbonhell** [comment]
Jesus christ yes

**Comment by Shdorsh** [comment]
It's been tested and is now ready

**Comment by Manly-Dwarf** [comment]
Jesus christ no
EDIT:
okay to prevent this from being a shitpost, no. ID locked weapons just sounds like a disaster, ID locked exploding drones sounds even more like a disaster. Circuits are already insanely annoying and easily exploitable, now you're going to gatekeep? Why?

**Comment by Carbonhell** [comment]
Id locked doesn't mean only the id owner can use the circuit, but rather that the circuit code cannot be scanned

**Comment by Manly-Dwarf** [comment]
So unnecessary gatekeeping to encourage "creativity" and "experimentation" even though 90% of this shit just comes from a discord. Okay.

**Comment by Shdorsh** [comment]
it enables people to learn how to use circuits themselves instead of copypasting it, which will have its nerf to be only accessible a slight bit later in round on my next pr

**Comment by Carbonhell** [comment]
what's the point of that
If the code is public you can easily get access to it too
plus idlocking only stops the analyzer from working,if it is only idlocked you can simply check it normally and see what circuits it's made of

**Comment by Manly-Dwarf** [comment]
Is copypasting even a bad thing, really? The creative people who take the time to learn circuits will be able to do more than the copypastes, whereas those who don't can only use the machines they've been told about.
I fail to see why this is a problem or a necessary inclusion. It just seems like a way to make circuitry an even bigger secret club and circlejerk.
Maybe I'm just a brick wall when it comes to circuits and I will hate them eternally for smoke & stun circuits, I don't know. But this seems like a great way to keep people _out_ of circuits, which isn't a good thing.

**Comment by Shdorsh** [comment]
My theory is: copypasting isn't bad, copypasting instantly is

File: hippiestation/code/modules/integrated_electronics/core/assemblies.dm [hunk 1]
```diff
@@ -540,9 +547,60 @@
...
 	if(can_anchor && default_unfasten_wrench(user, I, 20))
 		return
+
+	// ID-Lock part: check if we have an id-lock and only lock if we're not trying to get values from it, to prevent accidents
+	if(istype(I, /obj/item/integrated_electronics/debugger))
+		var/obj/item/integrated_electronics/debugger/debugger = I
+		if(debugger.idlock)
+			// check if unlocked to lock
+			if(!idlock)
+				idlock = debugger.idlock
```
Comment (Carbonhell) [inline hippiestation/code/modules/integrated_electronics/core/assemblies.dm:604]: This being a ref can cause GC problems and i think it is not the proper way to implement this idea
A better one would be a simple locking circuit that locks/unlocks the assembly when pulsed,let the user choose how (keypad,id,etc)

**Review by Carbonhell — CHANGES_REQUESTED** [review]

**Comment by Manly-Dwarf** [comment]
Now, I'm not against that, but
>Make it so that greyshits don't steal other people's design and have to come up with their own. Also, make circuits more secure to use so people can show off their best without the constant fear of the circuit analyzer.

If you want to just add a typing delay or something that'll slow down the rate of copypasting circuits, then whatever I guess?
But this is clearly not the intended goal.

**Comment by Shdorsh** [comment]
the fact that you do it instantaneously gets you direct access to what would have else been a few minutes of work. making it available later in the round by needing more research points would be nice. also, a circuit club? you can find some code on various discords that lets you learn the basics

**Comment by Manly-Dwarf** [comment]
Let the people have the circuits, not the club! We SEIZE your circuit properties, damn it!

**Comment by Shdorsh** [comment]
it's only closed source if the people decide to id-lock, even with the pr implemented. if people had the intent to go around sharing cancer to ruin the station, nothing on this pr would stop them, i'm merely giving them the possibility to not do so.

**Comment by Manly-Dwarf** [comment]
And I say that possibility shouldn't be there at all. If you're going to make stupid machines, you run the "risk" of other people using them as well. This goes for pretty much anything in the game. Atmos is one such good example, where you can't exactly "hide" all of the pipes running around.

**Comment by Shdorsh** [comment]
also, the slowed down copying will be another pr. the instant cloning disk will still be available, there will just be another rapid cloning disk, that takes the instant one's current place and the instant will cost a few more research points

**Comment by Shdorsh** [comment]
if you know how it works and learned some circuitry, you should be able to recreate it yourself, no? well, we'll see what the greater audience has to say on this issue.

**Comment by DaydreamPuppeh** [comment]
Honestly, I think this will make people use circuits less often. -1

**Comment by ghost** [comment]
+1 some circuits are too strong for some greytide to just scan and use forever

**Comment by Carbonhell** [comment]
@HideAndSeekLOGIC this pr is not aimed at preventing that, it's to allow nice big circuits to be done and people to always know who did it

If a circuit is overpowered it must be nerfed

**Comment by DaydreamPuppeh** [comment]
Switching to a +1. I misunderstood the point of this

**Comment by Carbonhell** [comment]
Tbh idlocking sounds fine,it pushes people towards actually understanding a circuit rather than copypasting code. But i cannot help but feel that sealing may be a bit too op,considering there is no counter anyhow.

**Comment by Shdorsh** [comment]
There would be a way if we made the circuit reopenable but have a chance to destroy pin connections. I'd make it roll a prob(30) chance for each circuit in the assembly to remove all its connections.

**Comment by Manly-Dwarf** [comment]
free shrug
Honestly I'm just in the party that hates Circuits' entire existence.

>this pr is not aimed at preventing that, it's to allow nice big circuits to be done and people to always know who did it

Encouraging meta through code, are we Carbon?

**Comment by Carbonhell** [comment]
what does that mean, meta is unrelated to this,it has no effect whatsoever
But say i do some sort of big circuit to help medics,  but i want to keep the code private to avoid it being replicated and filled with some bad chems, this PR would help me
Your reasonings apply for sealing, i agree it is a bit too op right now, but id locking is fine really, it pushes people interested in how a circuit works to actually use their heads and see how it works

**Comment by ghost** [comment]
make people unable to directly get the "source code" of the circuit, but let them unseal the circuit and manually see how it works

**Comment by Shdorsh** [comment]
@zamolxius that's just what id-lock does: turns source code ungettable. however, that would mean that people could still just take the circuit, open it up and get a general understanding of it. that's why there's sealing.

**Comment by Shdorsh** [comment]
From the feedback judging, seems like people who use circuitry are alright with it. I'll just quickly do something with weakrefs and tweak sealing a bit and then it'll be done.

File: hippiestation/code/modules/integrated_electronics/core/assemblies.dm (hunk 1 above)
Comment (Shdorsh) [reply hippiestation/code/modules/integrated_electronics/core/assemblies.dm:604]: resolved due to usage of weakref (it's in the debugger's code: stores all the refs to weakrefs). just gimme some time to fix the printers so they do the same

**Comment by Shdorsh** [comment]
It's done.

**Comment by DeltaFire15** [comment]
What about emags being used on the circuit?

**Comment by Shdorsh** [comment]
No emagging circuits currently in existence.

**Comment by Carbonhell** [comment]
Idlocking is fine, but sealing may be too op. If you unify both, you have a total closedsource bot and that is definitely bad, as goyon said. Either make downsides for the sealing process (for example, you can unseal with a welder but it breaks random pins and such, perhaps), or perhaps let people choose between one or the other, but not both, not sure.
Once the required nerf is in, +1

**Comment by Shdorsh** [comment]
@Carbonhell that's what I did on the last update, prob(50) to remove the pin connections for each circuit in the assembly.

**Comment by earthmoonjune** [comment]
+1

**Review by GuyonBroadway — APPROVED** [review]
Well this looks ok I think.

**Review by Carbonhell — APPROVED** [review]

**Pull request merged**